<a href="https://colab.research.google.com/github/GDevigili/TCC-DataWarehouse/blob/load-ibge-localidades/src/localidades.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import requests
import json
import sys

sys.path.append('..')

# Data Manipulation
import pandas as pd

from utils import *

In [2]:
# # Run this cells if running in google colab
# from google.colab import data_table
# data_table.enable_dataframe_formatter()

# Data Loading

In [3]:
url = "https://servicodados.ibge.gov.br/api/v1/localidades/distritos"
request = get_request().get(url)
info = request.json()

Info é uma lista de dicionários que vem de `request.json()`

### Sample data

```py
{
  'id': 520005005,
  'nome': 'Abadia de Goiás',
  'municipio': {
    'id': 5200050,
    'nome': 'Abadia de Goiás',
    'microrregiao': {
      'id': 52010,
      'nome': 'Goiânia',
      'mesorregiao': {
        'id': 5203,
        'nome': 'Centro Goiano',
        'UF': {
          'id': 52,
          'sigla': 'GO',
          'nome': 'Goiás',
          'regiao': {
            'id': 5,
            'sigla': 'CO',
            'nome': 'Centro-Oeste'
          }
        }
      }
    },
    'regiao-imediata': {
      'id': 520001,
      'nome': 'Goiânia',
      'regiao-intermediaria': {
        'id': 5201,
        'nome': 'Goiânia',
        'UF': {
          'id': 52,
          'sigla': 'GO',
          'nome': 'Goiás',
          'regiao': {
            'id': 5,
            'sigla': 'CO',
            'nome': 'Centro-Oeste'
          }
        }
      }
    }
  }
}
```

In [4]:
m_ = info[0]
info[0]

{'id': 520005005,
 'nome': 'Abadia de Goiás',
 'municipio': {'id': 5200050,
  'nome': 'Abadia de Goiás',
  'microrregiao': {'id': 52010,
   'nome': 'Goiânia',
   'mesorregiao': {'id': 5203,
    'nome': 'Centro Goiano',
    'UF': {'id': 52,
     'sigla': 'GO',
     'nome': 'Goiás',
     'regiao': {'id': 5, 'sigla': 'CO', 'nome': 'Centro-Oeste'}}}},
  'regiao-imediata': {'id': 520001,
   'nome': 'Goiânia',
   'regiao-intermediaria': {'id': 5201,
    'nome': 'Goiânia',
    'UF': {'id': 52,
     'sigla': 'GO',
     'nome': 'Goiás',
     'regiao': {'id': 5, 'sigla': 'CO', 'nome': 'Centro-Oeste'}}}}}}

Existem localidades repetidas, necessário limpeza disso

In [5]:
for i in range(5):
  print(info[i]['municipio'])

{'id': 5200050, 'nome': 'Abadia de Goiás', 'microrregiao': {'id': 52010, 'nome': 'Goiânia', 'mesorregiao': {'id': 5203, 'nome': 'Centro Goiano', 'UF': {'id': 52, 'sigla': 'GO', 'nome': 'Goiás', 'regiao': {'id': 5, 'sigla': 'CO', 'nome': 'Centro-Oeste'}}}}, 'regiao-imediata': {'id': 520001, 'nome': 'Goiânia', 'regiao-intermediaria': {'id': 5201, 'nome': 'Goiânia', 'UF': {'id': 52, 'sigla': 'GO', 'nome': 'Goiás', 'regiao': {'id': 5, 'sigla': 'CO', 'nome': 'Centro-Oeste'}}}}}
{'id': 3100104, 'nome': 'Abadia dos Dourados', 'microrregiao': {'id': 31019, 'nome': 'Patrocínio', 'mesorregiao': {'id': 3105, 'nome': 'Triângulo Mineiro/Alto Paranaíba', 'UF': {'id': 31, 'sigla': 'MG', 'nome': 'Minas Gerais', 'regiao': {'id': 3, 'sigla': 'SE', 'nome': 'Sudeste'}}}}, 'regiao-imediata': {'id': 310061, 'nome': 'Monte Carmelo', 'regiao-intermediaria': {'id': 3111, 'nome': 'Uberlândia', 'UF': {'id': 31, 'sigla': 'MG', 'nome': 'Minas Gerais', 'regiao': {'id': 3, 'sigla': 'SE', 'nome': 'Sudeste'}}}}}
{'id'

# Preparing the Dataset

## Data Unpacking

In [6]:
# def unpack_json(json_dict: dict, row: dict = {}, parent_col = '') -> dict:
#   """
#     Make a documentation (how I wish colab had copilot :/ )
#   """
#   # Iterate over each key/value pair on the data
#   for key in json_dict:
#     # If the value is another dict, i.e. it has more data
#     if type(json_dict[key]) == dict:
#       # recursivelly call unpack_json() to extract the data
#       # passing the a json_dict, the current row (pointer), and the name of the parent col
#       unpack_json(json_dict[key], row = row, parent_col = key)
#     else:
#       # If it's a value (e.g. str, int, etc.), register it on the row
#       # formatting the row's column name
#       if parent_col == '':
#         col_name = str(parent_col) + str(key)
#       else:
#         col_name = str(parent_col) + '_' + str(key)
#       # register the value in the row
#       row[col_name] = json_dict[key]
#   return row.copy()

In [7]:
rows = []
for i in range(len(info)):
  data = unpack_json(info[i])
  rows.append(data)

## Making a Data Frame

In [8]:
df = pd.DataFrame(rows)

In [9]:
df

id                 nome  municipio-id       municipio-nome  \
0      520005005      Abadia de Goiás       5200050      Abadia de Goiás   
1      310010405  Abadia dos Dourados       3100104  Abadia dos Dourados   
2      520010005            Abadiânia       5200100            Abadiânia   
3      520010010       Posse d'Abadia       5200100            Abadiânia   
4      310020305               Abaeté       3100203               Abaeté   
...          ...                  ...           ...                  ...   
10665  293360415               Iguira       2933604          Xique-Xique   
10666  251740705               Zabelê       2517407               Zabelê   
10667  355715405             Zacarias       3557154             Zacarias   
10668  211400705              Zé Doca       2114007              Zé Doca   
10669  421985305               Zortéa       4219853               Zortéa   

       microrregiao-id      microrregiao-nome  mesorregiao-id  \
0                52010                Goiânia            5203   
1                31019             Patrocínio            3105   
2                52012    Entorno de Brasília            5204   
3                52012    Entorno de Brasília            5204   
4                31024            Três Marias            3106   
...                ...                    ...             ...   
10665            29006                  Barra            2902   
10666            25010       Cariri Ocidental            2502   
10667            35004  São José do Rio Preto            3501   
10668            21008                Pindaré            2102   
10669            42009            Curitibanos            4203   

                       mesorregiao-nome  UF-id UF-sigla         UF-nome  \
0                         Centro Goiano     52       GO           Goiás   
1      Triângulo Mineiro/Alto Paranaíba     31       MG    Minas Gerais   
2                          Leste Goiano     52       GO           Goiás   
3                          Leste Goiano     52       GO           Goiás   
4                       Central Mineira     31       MG    Minas Gerais   
...                                 ...    ...      ...             ...   
10665     Vale São-Franciscano da Bahia     29       BA           Bahia   
10666                         Borborema     25       PB         Paraíba   
10667             São José do Rio Preto     35       SP       São Paulo   
10668                  Oeste Maranhense     21       MA        Maranhão   
10669                           Serrana     42       SC  Santa Catarina   

       regiao-id regiao-sigla   regiao-nome  regiao-imediata-id  \
0              5           CO  Centro-Oeste              520001   
1              3           SE       Sudeste              310061   
2              5           CO  Centro-Oeste              520002   
3              5           CO  Centro-Oeste              520002   
4              3           SE       Sudeste              310070   
...          ...          ...           ...                 ...   
10665          2           NE      Nordeste              290021   
10666          2           NE      Nordeste              250007   
10667          3           SE       Sudeste              350023   
10668          2           NE      Nordeste              210009   
10669          4            S           Sul              420008   

           regiao-imediata-nome  regiao-intermediaria-id  \
0                       Goiânia                     5201   
1                 Monte Carmelo                     3111   
2                      Anápolis                     5201   
3                      Anápolis                     5201   
4                        Abaeté                     3113   
...                         ...                      ...   
10665       Xique-Xique - Barra                     2907   
10666                  Monteiro                     2502   
10667       Birigui - Penápolis                     3506   
10668         

# Data Exploration

Showing a sample of the data:

In [10]:
df

id                 nome  municipio-id       municipio-nome  \
0      520005005      Abadia de Goiás       5200050      Abadia de Goiás   
1      310010405  Abadia dos Dourados       3100104  Abadia dos Dourados   
2      520010005            Abadiânia       5200100            Abadiânia   
3      520010010       Posse d'Abadia       5200100            Abadiânia   
4      310020305               Abaeté       3100203               Abaeté   
...          ...                  ...           ...                  ...   
10665  293360415               Iguira       2933604          Xique-Xique   
10666  251740705               Zabelê       2517407               Zabelê   
10667  355715405             Zacarias       3557154             Zacarias   
10668  211400705              Zé Doca       2114007              Zé Doca   
10669  421985305               Zortéa       4219853               Zortéa   

       microrregiao-id      microrregiao-nome  mesorregiao-id  \
0                52010                Goiânia            5203   
1                31019             Patrocínio            3105   
2                52012    Entorno de Brasília            5204   
3                52012    Entorno de Brasília            5204   
4                31024            Três Marias            3106   
...                ...                    ...             ...   
10665            29006                  Barra            2902   
10666            25010       Cariri Ocidental            2502   
10667            35004  São José do Rio Preto            3501   
10668            21008                Pindaré            2102   
10669            42009            Curitibanos            4203   

                       mesorregiao-nome  UF-id UF-sigla         UF-nome  \
0                         Centro Goiano     52       GO           Goiás   
1      Triângulo Mineiro/Alto Paranaíba     31       MG    Minas Gerais   
2                          Leste Goiano     52       GO           Goiás   
3                          Leste Goiano     52       GO           Goiás   
4                       Central Mineira     31       MG    Minas Gerais   
...                                 ...    ...      ...             ...   
10665     Vale São-Franciscano da Bahia     29       BA           Bahia   
10666                         Borborema     25       PB         Paraíba   
10667             São José do Rio Preto     35       SP       São Paulo   
10668                  Oeste Maranhense     21       MA        Maranhão   
10669                           Serrana     42       SC  Santa Catarina   

       regiao-id regiao-sigla   regiao-nome  regiao-imediata-id  \
0              5           CO  Centro-Oeste              520001   
1              3           SE       Sudeste              310061   
2              5           CO  Centro-Oeste              520002   
3              5           CO  Centro-Oeste              520002   
4              3           SE       Sudeste              310070   
...          ...          ...           ...                 ...   
10665          2           NE      Nordeste              290021   
10666          2           NE      Nordeste              250007   
10667          3           SE       Sudeste              350023   
10668          2           NE      Nordeste              210009   
10669          4            S           Sul              420008   

           regiao-imediata-nome  regiao-intermediaria-id  \
0                       Goiânia                     5201   
1                 Monte Carmelo                     3111   
2                      Anápolis                     5201   
3                      Anápolis                     5201   
4                        Abaeté                     3113   
...                         ...                      ...   
10665       Xique-Xique - Barra                     2907   
10666                  Monteiro                     2502   
10667       Birigui - Penápolis                     3506   
10668         

In [11]:
df.info(show_counts = True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10670 entries, 0 to 10669
Data columns (total 18 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   id                         10670 non-null  int64 
 1   nome                       10670 non-null  object
 2   municipio-id               10670 non-null  int64 
 3   municipio-nome             10670 non-null  object
 4   microrregiao-id            10670 non-null  int64 
 5   microrregiao-nome          10670 non-null  object
 6   mesorregiao-id             10670 non-null  int64 
 7   mesorregiao-nome           10670 non-null  object
 8   UF-id                      10670 non-null  int64 
 9   UF-sigla                   10670 non-null  object
 10  UF-nome                    10670 non-null  object
 11  regiao-id                  10670 non-null  int64 
 12  regiao-sigla               10670 non-null  object
 13  regiao-nome                10670 non-null  object
 14  regiao

#### Unique counts

In [12]:
cols_info = []
for column in df.columns:
  cols_info.append({
      'col_name': column,
      'unique_count': len(df[column].unique())
  })
  
cols_info

[{'col_name': 'id', 'unique_count': 10670},
 {'col_name': 'nome', 'unique_count': 9193},
 {'col_name': 'municipio-id', 'unique_count': 5570},
 {'col_name': 'municipio-nome', 'unique_count': 5297},
 {'col_name': 'microrregiao-id', 'unique_count': 558},
 {'col_name': 'microrregiao-nome', 'unique_count': 554},
 {'col_name': 'mesorregiao-id', 'unique_count': 137},
 {'col_name': 'mesorregiao-nome', 'unique_count': 137},
 {'col_name': 'UF-id', 'unique_count': 27},
 {'col_name': 'UF-sigla', 'unique_count': 27},
 {'col_name': 'UF-nome', 'unique_count': 27},
 {'col_name': 'regiao-id', 'unique_count': 5},
 {'col_name': 'regiao-sigla', 'unique_count': 5},
 {'col_name': 'regiao-nome', 'unique_count': 5},
 {'col_name': 'regiao-imediata-id', 'unique_count': 510},
 {'col_name': 'regiao-imediata-nome', 'unique_count': 508},
 {'col_name': 'regiao-intermediaria-id', 'unique_count': 133},
 {'col_name': 'regiao-intermediaria-nome', 'unique_count': 133}]


**OBSERVAÇÕES PARA FUTURA LIMPEZA DE DADOS:**


Existem municípios com nome repetido mas ID distintos, possibilidades:
+ Município localizado em duas micro/meso regiões/UFs distintas
+ Municípios com mesmo nome

(O mesmo se propaga para demais dados.)

Existem 5568 municípios + 2 distritos (federal e Fernando de Noronha) que compõem os 5570 id's de município

# Other Datasets

## Regiões Integradas de Desenvolvimento

In [24]:
url = 'https://servicodados.ibge.gov.br/api/v1/localidades/regioes-integradas-de-desenvolvimento'

In [25]:
df_regioes_desenv = make_df(get_request_json(url))
df_regioes_desenv

id                                               nome  municipio-id  \
0  01301  Região Integrada de Desenvolvimento da Grande ...       3205309   
1  03101  Região Administrativa Integrada de Desenvolvim...       3205309   
2  07801  Região Integrada de Desenvolvimento do Distrit...       3205309   

  municipio-nome  microrregiao-id microrregiao-nome  mesorregiao-id  \
0        Vitória            32009           Vitória            3203   
1        Vitória            32009           Vitória            3203   
2        Vitória            32009           Vitória            3203   

            mesorregiao-nome  UF-id UF-sigla  ... regiao-id  regiao-sigla  \
0  Central Espírito-santense     32       ES  ...         3            SE   
1  Central Espírito-santense     32       ES  ...         3            SE   
2  Central Espírito-santense     32       ES  ...         3            SE   

  regiao-nome regiao-imediata-id  regiao-imediata-nome  \
0     Sudeste             320001               Vitória   
1     Sudeste             320001               Vitória   
2     Sudeste             320001               Vitória   

  regiao-intermediaria-id  regiao-intermediaria-nome  \
0                    3201                    Vitória   
1                    3201                    Vitória   
2                    3201                    Vitória   

                                          municipios distrito-id  \
0  [{'id': 2112209, 'nome': 'Timon', 'UF': {'id':...   320530910   
1  [{'id': 2608750, 'nome': 'Lagoa Grande', 'UF':...   320530910   
2  [{'id': 3104502, 'nome': 'Arinos', 'UF': {'id'...   320530910   

   distrito-nome  
0     Goiabeiras  
1     Goiabeiras  
2     Goiabeiras  

[3 rows x 21 columns]

## Subdistritos

In [26]:
url = 'https://servicodados.ibge.gov.br/api/v1/localidades/subdistritos'

In [27]:
df_subdistritos = make_df(get_request_json(url))
df_subdistritos

id                                         nome  municipio-id  \
0    43004060506                                 Zona Central       4300406   
1    43004060507               Zona da Coxilha ou Cidade Alta       4300406   
2    43004060508                                   Zona Norte       4300406   
3    43004060509                                     Zona Sul       4300406   
4    43004060510                                   Zona Leste       4300406   
..           ...                                          ...           ...   
638  32053090510     Região Administrativa V - Praia do Canto       3205309   
639  32053090512        Região Administrativa VII - São Pedro       3205309   
640  32053091013        Região Administrativa VI - Goiabeiras       3205309   
641  32053091014  Região Administrativa VIII - Jardim Camburi       3205309   
642  32053091015   Região Administrativa IX - Jardim da Penha       3205309   

    municipio-nome  microrregiao-id   microrregiao-nome  mesorregiao-id  \
0         Alegrete            43029  Campanha Ocidental            4306   
1         Alegrete            43029  Campanha Ocidental            4306   
2         Alegrete            43029  Campanha Ocidental            4306   
3         Alegrete            43029  Campanha Ocidental            4306   
4         Alegrete            43029  Campanha Ocidental            4306   
..             ...              ...                 ...             ...   
638        Vitória            32009             Vitória            3203   
639        Vitória            32009             Vitória            3203   
640        Vitória            32009             Vitória            3203   
641        Vitória            32009             Vitória            3203   
642        Vitória            32009             Vitória            3203   

              mesorregiao-nome  UF-id UF-sigla  ... regiao-id  regiao-sigla  \
0       Sudoeste Rio-grandense     43       RS  ...         4             S   
1       Sudoeste Rio-grandense     43       RS  ...         4             S   
2       Sudoeste Rio-grandense     43       RS  ...         4             S   
3       Sudoeste Rio-grandense     43       RS  ...         4             S   
4       Sudoeste Rio-grandense     43       RS  ...         4             S   
..                         ...    ...      ...  ...       ...           ...   
638  Central Espírito-santense     32       ES  ...         3            SE   
639  Central Espírito-santense     32       ES  ...         3            SE   
640  Central Espírito-santense     32       ES  ...         3            SE   
641  Central Espírito-santense     32       ES  ...         3            SE   
642  Central Espírito-santense     32       ES  ...         3            SE   

    regiao-nome regiao-imediata-id  regiao-imediata-nome  \
0           Sul             430015            Uruguaiana   
1           Sul             430015            Uruguaiana   
2           Sul             430015            Uruguaiana   
3           Sul             430015            Uruguaiana   
4           Sul             430015            Uruguaiana   
..          ...                ...                   ...   
638     Sudeste             320001               Vitória   
639     Sudeste             320001               Vitória   
640     Sudeste             320001               Vitória   
641     Sudeste             320001               Vitória   
642     Sudeste             320001               Vitória   

    regiao-intermediaria-id  regiao-intermediaria-nome  \
0                      4304                 Uruguaiana   
1                      4304                 Uruguaiana   
2                      4304                 Uruguaiana   
3                      4304                 Uruguaiana   
4                      4304                 Uruguaiana   
..                      ...                        ...   
638                    3201                    Vitória   
639                    3201   

In [29]:
df_subdistritos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 643 entries, 0 to 642
Data columns (total 21 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   id                         643 non-null    int64 
 1   nome                       643 non-null    object
 2   municipio-id               643 non-null    int64 
 3   municipio-nome             643 non-null    object
 4   microrregiao-id            643 non-null    int64 
 5   microrregiao-nome          643 non-null    object
 6   mesorregiao-id             643 non-null    int64 
 7   mesorregiao-nome           643 non-null    object
 8   UF-id                      643 non-null    int64 
 9   UF-sigla                   643 non-null    object
 10  UF-nome                    643 non-null    object
 11  regiao-id                  643 non-null    int64 
 12  regiao-sigla               643 non-null    object
 13  regiao-nome                643 non-null    object
 14  regiao-ime

## Regiões Metropolitanas

In [31]:
url = 'https://servicodados.ibge.gov.br/api/v1/localidades/regioes-metropolitanas/'

In [32]:
df_reg_metropolitanas = make_df(get_request_json(url))
df_reg_metropolitanas

id                                            nome  municipio-id  \
0   00101             Região Metropolitana de Porto Velho       3205309   
1   00201                  Região Metropolitana de Manaus       3205309   
2   00301                 Região Metropolitana da Capital       3205309   
3   00401                 Região Metropolitana de Central       3205309   
4   00501           Região Metropolitana do Sul do Estado       3205309   
..    ...                                             ...           ...   
79  07401            Região Metropolitana de Porto Alegre       3205309   
80  07501            Região Metropolitana da Serra Gaúcha       3205309   
81  07601      Região Metropolitana do Vale do Rio Cuiabá       3205309   
82  07602  Entorno Metropolitano da RM Vale do Rio Cuiabá       3205309   
83  07701                 Região Metropolitana de Goiânia       3205309   

   municipio-nome  microrregiao-id microrregiao-nome  mesorregiao-id  \
0         Vitória            32009           Vitória            3203   
1         Vitória            32009           Vitória            3203   
2         Vitória            32009           Vitória            3203   
3         Vitória            32009           Vitória            3203   
4         Vitória            32009           Vitória            3203   
..            ...              ...               ...             ...   
79        Vitória            32009           Vitória            3203   
80        Vitória            32009           Vitória            3203   
81        Vitória            32009           Vitória            3203   
82        Vitória            32009           Vitória            3203   
83        Vitória            32009           Vitória            3203   

             mesorregiao-nome  UF-id UF-sigla  ... regiao-sigla   regiao-nome  \
0   Central Espírito-santense     11       RO  ...            N         Norte   
1   Central Espírito-santense     13       AM  ...            N         Norte   
2   Central Espírito-santense     14       RR  ...            N         Norte   
3   Central Espírito-santense     14       RR  ...            N         Norte   
4   Central Espírito-santense     14       RR  ...            N         Norte   
..                        ...    ...      ...  ...          ...           ...   
79  Central Espírito-santense     43       RS  ...            S           Sul   
80  Central Espírito-santense     43       RS  ...            S           Sul   
81  Central Espírito-santense     51       MT  ...           CO  Centro-Oeste   
82  Central Espírito-santense     51       MT  ...           CO  Centro-Oeste   
83  Central Espírito-santense     52       GO  ...           CO  Centro-Oeste   

   regiao-imediata-id regiao-imediata-nome  regiao-intermediaria-id  \
0              320001              Vitória                     3201   
1              320001              Vitória                     3201   
2              320001              Vitória                     3201   
3              320001              Vitória                     3201   
4              320001              Vitória                     3201   
..                ...                  ...                      ...   
79             320001              Vitória                     3201   
80             320001              Vitória                     3201   
81             320001              Vitória                     3201   
82             320001              Vitória                     3201   
83             320001              Vitória                     3201   

   regiao-intermediaria-nome  \
0                    Vitória   
1                    Vitória   
2                    Vitória   
3                    Vitória   
4                    Vitória   
..                       ...   
79                   Vitória   
80                   Vitória   
81                   Vitória   
82                   Vitória   
83                   Vitória   

                              